persistence/aggregate.go

```diff
@@ -32,6 +35,21 @@ type AggregateMetaData struct {
 	BarrierEventID string
 }
 
+type AggregateSnapshot struct {
+	// State is the instance's current aggregate state.
+	State dogma.AggregateRoot
```

Comment:
> I'm a little unsure about how to best capture the current state for the snapshot. AFAIK, dogma.AggregateRoot needs to be persisted, but this may be incorrect.

Yep, this is correct. 

The [`persistence.ProcessInstance`](https://github.com/dogmatiq/verity/blob/main/persistence/process.go#L10) type might serve as a decent reference here. It's used to persist `dogma.ProcessRoot` instances, which is a pretty similar idea.

Note the use of [`marshalkit.Packet`](https://pkg.go.dev/github.com/dogmatiq/marshalkit#Packet). All user-defined data types (messages, aggregate roots, process roots, etc) are persisted using this structure. The upshot is that the `verity/persistence` package remains unaware of the user-defined types, it just operates on binary representations.

```suggestion
	// Packet contains the binary representation of the aggregate snapshot.
	Packet marshalkit.Packet
```

(Pedantic: for consistency with the other data types, I'd also move this field to the end of the struct).

Reply:
See also [`verity.NewDefaultMarshaler()`](https://github.com/dogmatiq/verity/blob/b3ddea0ef27b300172bf33911ec64b807b20c4e7/engineoption.go#L181) which constructs a [`marshalkit.Marshaler`](https://pkg.go.dev/github.com/dogmatiq/marshalkit#Marshaler) to use for a given `dogma.Application` instance.  This will need to be modified to make the marshaler aware of the aggregate root types, again similar to the way the process roots work.

Reply:
@jmalloc 

RE:

> See also verity.NewDefaultMarshaler() which constructs a marshalkit.Marshaler to use for a given dogma.Application instance. This will need to be modified to make the marshaler aware of the aggregate root types, again similar to the way the process roots work.

Could you elaborate on this?

Are you suggesting that a new Marshaler needs to be created to accept unmarshaling of `dogma.AggregateRoot` type?

IIUC, `dogma.ProcessRoot` can be unmarshalled due to its type being a marker `interface{}` with no methods, where as `dogma.AggregateRoot` defines an `ApplyEvent` method, which results in a compile error when attempting to unmarshal to this type using the same process. 

See https://github.com/dogmatiq/verity/blob/main/handler/process/loader.go#L53

Reply:
> Are you suggesting that a new Marshaler needs to be created to accept unmarshaling of dogma.AggregateRoot type?

Not a new marshaler no, I just mean that the `NewDefaultMarshaler()` function in Verity has to be modified to make the marshaler it constructs aware of the aggregate types.

It's already aware of the process types because of this code here: https://github.com/dogmatiq/verity/blob/b3ddea0ef27b300172bf33911ec64b807b20c4e7/engineoption.go#L188